### PR TITLE
Force login/registration if not logged in

### PR DESCRIPTION
**Pull request by bbuchalter**
If an error occurs which prevents the registration or login process from
completing the assignment of "spree_current_user" but does complete the
assignment of "order.email", this would allow users to enter checkout.
However, checkout depends on "spree_current_user" being assigned; we do
not allow a guest checkout.

Therefore, we remove the conditional where the presence of "order.email"
would allow entry to checkout.

### CODE REVIEW

File: app/controllers/spree/checkout_controller_decorator.rb [hunk 1]
```diff
@@ -41,7 +41,7 @@ def check_authorization
     # Introduces a registration step whenever the +registration_step+ preference is true.
     def check_registration
       return unless Spree::Auth::Config[:registration_step]
-      return if spree_current_user or current_order.email
+      return if spree_current_user
```
Comment (gmacdougall) [inline app/controllers/spree/checkout_controller_decorator.rb]: I would prefer that this change to:

```Ruby
return if (
  spree_current_user ||
  (current_order.email && Spree::Config[:allow_guest_checkout])
)
```

Then we can cherry pick this change on to solidus_auth_devise in preparation for the future.

Comment (bbuchalter) [inline app/controllers/spree/checkout_controller_decorator.rb]: :+1: will change

**Comment by gmacdougall** [comment]
:+1:

If you have a chance could you please PR this in to `solidus_auth_devise`? I think it's an important fix. The maintainers over there will probably demand a test while you're at it.  :wink:

**Comment by bbuchalter** [comment]
Indeed.

**Pull request merged**